servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/From3Publisher.java

```diff
@@ -74,60 +78,34 @@ public void request(final long n) {
                 subscriber.onError(newExceptionForInvalidRequestN(n));
                 return;
             }
-            if (state == INIT) {
-                state = DELIVERED_V1;
-                try {
-                    subscriber.onNext(v1);
-                } catch (Throwable cause) {
-                    state = TERMINATED;
-                    subscriber.onError(cause);
-                    return;
-                }
-                // We could check CANCELLED here and return, but it isn't required.
-                if (n > 2) {
-                    deliverV2V3();
-                } else if (n == 2) {
-                    deliverV2();
-                }
-            } else if (state == DELIVERED_V1) {
-                if (n > 1) {
-                    deliverV2V3();
+            if (requested == 3) {
+                return;
+            }
+            requested = min(3, addWithOverflowProtection(requested, n));
+            while (delivered < requested) {
+                delivered++;
+                if (state == INIT) {
```

Comment:
I was also considering removing it, but then again 2-3 arguments will be the most common use-case so its probably worth the special version for it. Haven't done any perf testing though to see if the benefits are even visible.

Reply:
I run a quick JMH on the creation of the publisher, and the GC pressure is quite visible under the extremes of a benchmark. Still I think there is visible benefit for var-args < 5 arguments so I am leaning towards keeping the 3 args as well.

Reply:
Thinking about it from the other direction, if only `FromArrayPublisher` existed how likely would it be that it would appear as the bottleneck in someone's flame graph? We know that `FromXPublisher`s for X < 5 **could** be faster but are we sure it would make any difference outside of microbenchmarks?

An alternative implementation could also increase code reuse by abusing the initialization conditions so that From2Publisher and From3Publisher used the same code but the from-2 pattern initialized it such that value 1 was "already delivered".

Reply:
I generally agree with this line of thinking, and most of the times its not worth the effort and the duplication (potential more space for bugs). 

Having said that, _playing the devils advocate_, people do run benchmarks and they do compare us with alternative products, so its always good to not have to justify why we are generating more garbage. 
Also, outside the microbenchmarking world, I do thing generating less garbage is a good mental model, and in this case the gc rate was many folds bigger (10x); so if you use `Publisher#from` directly or indirectly (through zip) in your hot-path, you do pay for that additional cost.

clarification: there is no argument about speed here, its just gc pressure.
WDYT?

Reply:
We use `From3Publisher` in two places: `zip` operator for 3 singles and `HeaderUtils#setContentLength` to flatten the HTTP message. Let's keep this optimization. However, I like the suggestion to share code between `From2Publisher` and `From3Publisher`.

Reply:
Merged the two publishes in one. 👍